abi compatible data types feature request (#1375)
eclipse-score/score: docs/features/communication/abi_compatible_data_types/index.rst

Assumptions
^^^^^^^^^^^

* Shared memory regions are mapped at correctly aligned virtual addresses in both processes.
* No serialization or runtime copying occurs when interpreting a type from memory.
* Processes execute on the same instruction set architecture and use the same endianness.
* No synchronization or atomicity guarantees are defined at the data type level; these are provided by the IPC framework.
* All memory is allocated statically or pre-reserved. Dynamic memory allocation is disallowed.

Type Conformance
^^^^^^^^^^^^^^^^

Types used in shared memory must meet the following criteria:

1. **Fixed size and alignment**: Every type must have a known, constant size and alignment at compile time.
2. **Consistent layout across languages**: The layout of a type must be identical in Rust and C++ on the same platform.

> 3. **No pointers or references**: Types must not contain pointers to heap memory, function pointers, or references.

Reviewer: Changed to *"No absolute pointers or references"*.

4. **No language-specific metadata**: No vtables, slice headers, or implementation-specific type markers are allowed.

Each type definition must clearly indicate whether it conforms to these rules natively or requires a custom definition to do so.

Type Categories
^^^^^^^^^^^^^^^

Primitive Types
"""""""""""""""

These types are ABI-compatible when declared using fixed-size standard types:

.. list-table:: Native Type Mapping
   :header-rows: 1

   * - Concept
     - Rust
     - C++17
   * - Boolean
     - ``bool``
     - ``bool`` (1 byte)
   * - Integers (N = 8..128)
     - ``uN``, ``iN``
     - ``std::uintN_t``, ``std::intN_t``
   * - Floating point
     - ``f32``, ``f64``
     - ``float``, ``double``

All types must avoid trap representations and undefined padding.

Structs and Tuples
""""""""""""""""""

Structs and tuples are supported using standard layout rules:

* **Rust**: Requires ``#[repr(C)]``

> * **C++**: Requires ``standard_layout`` types with no inheritance or virtual functions

Reviewer 2: C++: standard_layout types are no virtual functions, and no 'virtual' inheritance, and only one class in the hierarchy has non-static data members.

Reviewer: Fixed in #1451

Motivation
==========

This feature request addresses specific challenges in achieving type compatibility within our inter-process communication (IPC) framework that leverages zero-copy shared memory mechanisms. Two essential scenarios are under evaluation:

1. **ABI Compatibility**: Processes implemented in different programming languages (C++17 and Rust 1.8x) must interpret a shared memory location consistently as the same native type, provided both have compile-time access to the type definition. This scenario eliminates serialization overhead and allows direct memory access.

2. **Type Description**: Processes responsible for translating data between internal ABI-compatible formats and external serialization formats should perform this translation without compile-time knowledge of type definitions. This runtime interpretation capability is critical for scalability, ensuring that gateway processes (or similar entities) can dynamically handle new data types without recompilation.

> 

Reviewer 3: I would explicitly include the record/replay use case. It means if the the IPC data is recorded, we'd need some sort of tooling (e.g. in Python) to process the data further. Additionally, some tooling for debugging would be required.

2. **Type Description**: Processes responsible for translating data between internal ABI-compatible formats and external serialization formats should perform this translation without compile-time knowledge of type definitions. This runtime interpretation capability is critical for scalability, ensuring that gateway processes and tools for recording/replaying/inspecting data can dynamically handle new data types without recompilation.

ABI Compatibility
-----------------

Our communication feature relies on shared memory to transfer data between processes. For effective zero-copy data exchange, processes written in C++17 and Rust 1.8x must inherently understand the data at shared memory locations identically. Achieving this requires ensuring that data types have consistent, fixed-size memory layouts.

This evaluation initially targets the following process configurations:

* Processes running on the same operating system.
* Processes running on different operating systems but under the same hypervisor.

Supporting different endianness between processes is explicitly out of scope, as it inherently demands bit manipulation, effectively requiring serialization.

> Different bit widths, however, are implicitly supported by specifying the width of all types and excluding word-size integers.

Reviewer 4: How would this play with alignments? I would imagine we end up in quite a huge mess if we have two parties accessing the same data but with a different bit width.

Maybe you mean that we want to support that our ABI compatible datatypes work on different bit widths, as long as all participants have the same bit width. If so, we should rephrase to make this more explicit.

> The following data types shall be supported by the IPC mechanism:

Reviewer 4: While IPC is a main driver, also crossing the FFI boundary without IPC is a driver. So let's not restrict this to IPC only.

```suggestion
The following data types shall be supported:
```

* **Primitive Types**:

  * Boolean
  * Numeric (fixed-size integers 8-128 bits, signed and unsigned; IEEE 754 floating-point numbers)

* **Sequence Types**:

  * Array (fixed-length)

* **User-Defined Types**:

  * Struct
  * Tuple
  * Enum (tag-only)
  * Variant ("tagged union"; *optional*)

* **Fixed-Size, Variable-Length Containers**:

  * Vector
  * Queue
  * Hash map (*optional*)
  * Hash set (*optional*)
  * Binary tree (*optional*)

* **Specialized Variants**:

  * Result
  * Option

All provided data types must ensure fixed size and consistent memory layouts.

Type Description
----------------

> A critical scalability feature involves gateway processes, which subscribe to IPC endpoints and translate ABI-compatible data types into external serialization formats. These gateways require the ability to interpret data without compile-time access to type definitions. To address this, an explicit runtime-readable type description format is necessary. This description allows dynamic, runtime interpretation of data structures, enabling the addition of new IPC topics without recompiling gateway processes.

Reviewer 4: See comment above on gateways and their transformations. I think reprocessing is a better driver for this.

In summary, the motivation behind this feature request is to define and standardize ABI-compatible data types and a runtime-accessible type description mechanism to ensure interoperability and scalability in zero-copy IPC scenarios involving multiple languages and dynamic environments.

.. Rationale
.. ==========

Specification
=============

ABI Compatibility
-----------------

This specification defines the set of rules and constraints for representing data types in shared memory such that they can be interpreted consistently across processes implemented in C++17 and Rust 1.8x. These types enable zero-copy inter-process communication by enforcing ABI compatibility at the memory layout level. The focus is on data exchange between processes using the same endianness.

Assumptions
^^^^^^^^^^^

* Shared memory regions are mapped at correctly aligned virtual addresses in both processes.
* No serialization or runtime copying occurs when interpreting a type from memory.
* Processes use the same endianness.

> * No synchronization or atomicity guarantees are defined at the data type level; these are provided by the IPC framework.

Reviewer 4: As mentioned above, FFI boundary crossing is also a use case. There user code must provide the synchronization.

```suggestion
* No synchronization or atomicity guarantees are defined at the data type level; these must be provided by the accessing code.
```

* All memory is allocated statically or pre-reserved. Dynamic memory allocation is disallowed.

Type Conformance
^^^^^^^^^^^^^^^^

Types used in shared memory must meet the following criteria:

1. **Fixed size and alignment**: Every type must have a known, constant size and alignment at compile time.
2. **Consistent layout across languages**: The layout of a type must be identical in Rust and C++, and on all platforms.

> 3. **No absolute pointers or references**: Types must not contain pointers to heap memory, function pointers, or references.

Reviewer 4: A pointer to stack is equally bad.
Also wording here is tricky. What we really want to say is that pointers and even references are fine, as long as they are implemented with a relative offset and do not point outside of the transferred data.

```suggestion
3. **No absolute pointers or references**: Types must not contain absolute pointers/function pointers/references that point out of the transferred data.
```

Reviewer 4: Sorry my "improved wording" is also wrong again...

```suggestion
3. **No absolute pointers or references**: Types must not contain absolute pointers/function pointers/references or any pointer/reference that point out of the transferred data.
```

4. **No language-specific metadata**: No vtables, slice headers, or implementation-specific type markers are allowed.

Each type definition must clearly indicate whether it conforms to these rules natively or requires a custom definition to do so.

Type Categories
^^^^^^^^^^^^^^^

Primitive Types
"""""""""""""""

These types are ABI-compatible when declared using fixed-size standard types:

.. list-table:: Native Type Mapping
   :header-rows: 1

   * - Concept
     - Rust
     - C++17
   * - Boolean
     - ``bool``
     - ``bool`` (1 byte, with ``0x00`` and ``0x01`` as the only valid bit patterns)
   * - Integers (N = 8, 16, 32, 64)
     - ``uN``, ``iN``
     - ``std::uintN_t``, ``std::intN_t``
   * - Floating point
     - ``f32``, ``f64``

>      - ``float``, ``double``

Reviewer 4: In C++ you would have to check that the floating point types adhere to IEEE754:

```c++
static_assert(std::numeric_limits<float>::is_iec559);
static_assert(std::numeric_limits<double>::is_iec559);
```

Furhter for floating point types we need to carefully consider whether it is fine that different FPUs work on the same data. FPUs are not standardized. This could lead to significant deviations in calculations.

Reviewer: I'll add a clarification that floating-point types must be IEEE 754 compliant.

I don't think we have to define the behavior of *operations* on those values; the feature request is concerned with data *exchange and storage*, not with data *processing*.

Reviewer 4: Hmm maybe we should just warn users that while the storage is compatible the operations may differ in the result.
This could lead to some very interesting bugs.

All types must avoid trap representations and undefined padding.

Structs and Tuples
""""""""""""""""""

Structs and tuples are supported using standard layout rules:

* **Rust**: Requires ``#[repr(C)]``
  (guarantees C-compatible memory layout;
  `full specification <https://doc.rust-lang.org/reference/type-layout.html#the-c-representation>`__)
* **C++**: Requires ``standard_layout``
  (no virtual functions, no virtual inheritance, and only one class in the hierarchy has non-static data members;
  `full specification <https://en.cppreference.com/w/cpp/language/classes.html#Standard-layout_class>`__)

Field ordering must be preserved and padding must be identical across compilers. Any alignment greater than the default must be explicitly declared.

Enums
"""""

Only fieldless enums with a defined underlying integer type are supported. These must use:

* ``#[repr(u8)]``, ``#[repr(u16)]``, etc. in Rust
* ``enum class MyEnum : std::uint8_t`` in C++

> Enums with payloads (discriminated unions) are not supported.

Reviewer 4: We could mention that a variant can be used to work around this limitation?

Not a must from my side, just an idea.

Arrays
""""""

Fixed-size arrays are naturally ABI-compatible and supported in both languages.

* Rust: ``[T; N]``
* C++: ``T[N]``

> Element types must also conform to this specification. No dynamic length information is allowed.

Reviewer 4: At least on C++ side we need something else for this. Using C-style arrays in C++ code is a slippery slope straight to hell.

I would go with something that is similar to a `std::array<>` but defines the ABI.
E.g. internally represented by a size field followed by the C-style array, but externally represented by a `std::array<>`-like interface.

Reviewer: You're right, this should be clarified. The idea is to have a `T[N]` memory layout, but a wrapper which checks for out-of-bounds accesses the same way Rust's `[T; N]` does.

   * - Structs
     - ✅ ``#[repr(C)]``
     - ✅ ``standard_layout``
     - Conforming
   * - Enums (fieldless)
     - ✅ ``#[repr(C)]``
     - ✅ With fixed base
     - Conforming
   * - Arrays
     - ✅ ``[T; N]``
     - ✅ ``T[N]``
     - Conforming
   * - Vector
     - ❌ ``Vec<T>``
     - ❌ ``std::vector<T>``
     - ✅ ``AbiVec<T, N>`` required
   * - Option
     - ❌ ``Option<T>``
     - ❌ ``std::optional<T>``
     - ✅ ``AbiOption<T>`` required
   * - Result
     - ❌ ``Result<T, E>``
     - ❌ ``std::expected<T, E>``
     - ✅ ``AbiResult<T, E>`` required

Type Description
----------------

To address the scenarios outlined in the motivation, a clearly defined type description mechanism is required. The type description provides sufficient information during runtime, enabling a process without compile-time access to type definitions to correctly interpret a given memory location according to the previously established ABI rules.

The goals are:

* Enable interpretation of shared memory content without compile-time access to type definitions.
* Support all ABI-compatible data types previously defined.
* Include versioning to manage schema evolution and compatibility.

> * Allow easy generation and parsing by tooling in both C++ and Rust.

Reviewer 4: Just as a side note, I personally think this is very hard to achieve with a type-safe, performant API and a static type system as in C++ and Rust.

**Definition-first Workflow**: Existing type definitions in Rust or C++ are the source of truth, and the corresponding type description is generated during the build process via compiler tooling.

Both workflows are valid, and the final decision is deferred pending further feasibility analysis.

Type Description Format
^^^^^^^^^^^^^^^^^^^^^^^

The format of the type description shall explicitly support versioning to allow schema evolution and backward compatibility. It must accommodate all data types described earlier in the ABI compatibility section. It should be simple, human-readable, and easily machine-parsable.

The choice of serialization format is left open but may include RON, JSON5, or a custom DSL, based on readability, tooling support, and maintainability.

.. Backwards Compatibility
.. =======================

Security Impact
===============

.. note::

   This section does not replace a formal security impact analysis. This only guides the design thoughts behind security related topics and is to be understood as a starting point for later workflows.

> No assumed impact on security.

Reviewer: I see, good point. How would you formulate this security aspect for the feature request?

Safety Impact
=============

.. note::

   This section does not replace a formal safety impact analysis. This only guides the design thoughts behind safety related topics and is to be understood as a starting point for later workflows.

All functionality shall be available to applications rated up to ASIL-B.

.. License Impact
.. ==============

.. How to Teach This
.. ==================

Rejected Ideas
==============

Reflection
----------

Reflection, in this context, is the ability to inspect data at runtime even if its structure is not or not fully known at compile time.
Benefits of reflection include being able to translate recorded data into a human-readable format (e.g., JSON or CSV) without having to know the type definitions at compile time; this enables general-purpose data recording and transformation tools.

This ability requires some form of *type description* being available at runtime, so that a sequence of bytes can be interpreted as a data structure.
There are two primary approaches to achieve this goal:

* *inline type descriptions*, which precede each instance of every type, and

> * *external type descriptions*, which are stored separately from the data.

Reviewer 4: I feel like you argue reflection a lot in the type description sections above. But here you reject the idea.
Could you rephrase the sections on type description to clear this up?
